File: src/rules/amount.py
"""金額規程チェック (§6.3 amount-rule check).

出張精算レポート (ExpenseReport) に対し, 金額面の妥当性を2系統で判定する.

(1) データ整合性 (data-integrity, 常時実行・規程不要):
    - 合計(申告) と 明細金額合計(計算) の一致
    - 行数(宣言) と 実明細数の一致
    - 各明細の 金額>=0 かつ 小計==金額

(2) 規程ベース上限判定 (rule-based, 旅費規定 必須):
    手当CD / account_name ごとに上限を引き, 役職・地域を考慮して判定.
    役職 (grade) 不明時は一般職上限以下→OK, 管理職上限以下→要確認, 超過→NG.
    旅費規定 未提供時は RULE_MISSING を返す.

(3) 勘定科目分類チェック:
    車(transport) かつ 貸借費(account_name) の場合は要確認.
"""
from __future__ import annotations

# 東京23区の区名リスト (都道府県フィールドに区名が含まれる場合に 東京23区 と判定)
_TOKYO_23_KU: frozenset[str] = frozenset([
    "千代田区", "中央区", "港区", "新宿区", "文京区",
    "台東区", "墨田区", "江東区", "品川区", "目黒区",
    "大田区", "世田谷区", "渋谷区", "中野区", "杉並区",
    "豊島区", "北区", "荒川区", "板橋区", "練馬区",
    "足立区", "葛飾区", "江戸川区",
])


def _is_tokyo_23ku(prefecture: str | None) -> bool:
    """都道府県フィールドが東京23区内かを判定する.

    prefecture に "東京" が含まれ、かつ23区のいずれかの区名が含まれる場合のみ True.
    "東京都八王子市" など多摩地区は False になる.
    prefecture が区名のみ ("港区") でも True を返す.
    """
    if not prefecture:
        return False
    pref = prefecture.strip()
    if "東京" in pref:
        return any(ku in pref for ku in _TOKYO_23_KU)
    return any(ku in pref for ku in _TOKYO_23_KU)


from models import (
    ExpenseReport,
    CheckResult,
    OK,
    NG,
    NEEDS_CHECK,
    RULE_MISSING,
    worst_status,
)
from config import Config


# ---------------------------------------------------------------------------
# データ整合性チェック
# ---------------------------------------------------------------------------

def _check_integrity(r: ExpenseReport) -> tuple[str, list[str], dict]:
    reasons: list[str] = []
    evidence: dict = {}
    statuses: list[str] = []

    gap = r.total_amount - r.computed_total
    has_allowance = any(
        (leg.allowance_cd_perdiem or leg.allowance_cd_lodging or leg.allowance_cd_stay)
        for leg in r.legs
    )
    if r.computed_total > r.total_amount:
        statuses.append(NG)
        reasons.append(
            f"明細金額合計が申告合計を超過: 計算{r.computed_total} > 申告{r.total_amount}"
        )
        evidence["合計超過"] = {
            "申告合計": r.total_amount,
            "計算合計": r.computed_total,
            "超過額": -gap,
        }
    elif gap > 0 and not has_allowance:
        statuses.append(NEEDS_CHECK)
        reasons.append(
            f"申告合計と明細合計に差額{gap}があるが手当コードなし: 要確認"
        )
        evidence["差額要確認"] = {
            "申告合計": r.total_amount,
            "計算合計": r.computed_total,
            "差額": gap,
        }
    elif gap > 0:
        evidence["手当相当差額"] = {
            "申告合計": r.total_amount,
            "明細金額合計": r.computed_total,
            "手当相当差額": gap,
        }

    if r.actual_leg_count != r.declared_leg_count:
        statuses.append(NEEDS_CHECK)
        reasons.append(
            f"行数不一致: 宣言{r.declared_leg_count} != 実{r.actual_leg_count}"
        )
        evidence["行数不一致"] = {
            "宣言行数": r.declared_leg_count,
            "実明細数": r.actual_leg_count,
        }

    bad_legs: list[dict] = []
    for leg in r.legs:
        if leg.amount < 0 or leg.subtotal != leg.amount:
            bad_legs.append({"明細No": leg.leg_no, "金額": leg.amount, "小計": leg.subtotal})
    if bad_legs:
        statuses.append(NG)
        reasons.append(f"明細金額異常 ({len(bad_legs)}件): 金額<0 または 小計!=金額")
        evidence["明細金額異常"] = bad_legs

    return worst_status(statuses), reasons, evidence


# ---------------------------------------------------------------------------
# 役職別上限ヘルパー
# ---------------------------------------------------------------------------

def _grade_caps(table: dict, grade: str | None) -> tuple[int | None, int | None]:
    """(exact_cap, None) if grade known; (一般職_cap, higher_cap) if grade unknown."""
    if not isinstance(table, dict):
        return None, None
    if grade:
        if grade in table:
            return table[grade], None
        # 主任以上エイリアス
        if any(t in grade for t in ("主任", "リーダー", "係長", "班長")):
            cap = table.get("主任以上")
            if cap is not None:
                return cap, None
        # 管理職エイリアス
        if any(t in grade for t in ("部長", "課長", "副参事", "参事", "取締役", "執行役")):
            cap = table.get("管理職")
            if cap is not None:
                return cap, None
        # フォールバック: 一般職
        cap = table.get("一般職")
        return cap, None
    # 役職不明: 一般職と最上位の差で判定
    cap_gen = table.get("一般職")
    cap_high = table.get("管理職") or table.get("主任以上")
    return cap_gen, cap_high


def _check_grade_allowance(
    leg_no: int, kind: str, amount: int,
    table: dict, grade: str | None,
    overs: list, needs_check: list,
) -> None:
    cap_exact, cap_high = _grade_caps(table, grade)

    if grade:
        # 役職確定: 単一上限で判定
        if cap_exact is None:
            needs_check.append({
                "明細No": leg_no, "種別": kind, "金額": amount,
                "理由": f"役職({grade})の上限が規程に未定義",
            })
        elif amount > cap_exact:
            overs.append({"明細No": leg_no, "種別": kind, "金額": amount,
                          "上限": cap_exact, "役職": grade})
    else:
        # 役職不明: 二段階判定
        if cap_exact is None:
            needs_check.append({"明細No": leg_no, "種別": kind, "金額": amount,
                                 "理由": "役職不明かつ規程未定義"})
            return
        if amount <= cap_exact:
            return  # 一般職上限以内 → OK
        if cap_high is not None and amount <= cap_high:
            needs_check.append({
                "明細No": leg_no, "種別": kind, "金額": amount,
                "一般職上限": cap_exact, "上位役職上限": cap_high,
                "理由": "役職確認が必要 (一般職超過・管理職上限以内)",
            })
        else:
            overs.append({
                "明細No": leg_no, "種別": kind, "金額": amount,
                "一般職上限": cap_exact,
                **({"上位役職上限": cap_high} if cap_high else {}),
                "理由": "全役職上限超過",
            })


# ---------------------------------------------------------------------------
# ホテル代チェック
# ---------------------------------------------------------------------------

def _check_hotel(
    leg_no: int, amount: int, prefecture: str | None, grade: str | None,
    hotel_limits: dict, overs: list, needs_check: list,
) -> None:
    region = "東京23区" if _is_tokyo_23ku(prefecture) else "その他"
    rtable = hotel_limits.get(region, hotel_limits.get("その他", {}))
    _check_grade_allowance(leg_no, f"ホテル代({region})", amount, rtable, grade, overs, needs_check)


# ---------------------------------------------------------------------------
# 長距離運転手当チェック
# ---------------------------------------------------------------------------

def _check_long_distance(leg_no: int, amount: int, km_lower: int | None,
                          has_lodging: bool, limits: dict,
                          overs: list, needs_check: list) -> None:
    ld = limits.get("長距離運転手当", {})
    if not ld:
        return
    trip_type = "宿泊時" if has_lodging else "日帰り"
    lower = ld.get(f"{trip_type}_下限")
    upper = ld.get(f"{trip_type}_上限")
    if lower is None or upper is None:
        return
    # 長距離加算
    add_km = ld.get("長距離加算_km", 300)
    add_amt = ld.get("長距離加算額", 1000)
    if km_lower is not None and km_lower >= add_km:
        upper = upper + add_amt
    if amount < lower:
        needs_check.append({
            "明細No": leg_no, "種別": "長距離運転手当", "金額": amount,
            "下限": lower, "上限": upper, "旅行種別": trip_type,
            "理由": "規程下限未満",
        })
    elif amount > upper:
        overs.append({
            "明細No": leg_no, "種別": "長距離運転手当", "金額": amount,
            "上限": upper, "旅行種別": trip_type,
        })


# ---------------------------------------------------------------------------
# 規程ベース上限判定
# ---------------------------------------------------------------------------

def _check_rules(r: ExpenseReport, cfg: Config) -> tuple[str, list[str], dict]:
    if not cfg.has_amount_rules():
        return (
            RULE_MISSING,
            ["旅費規定 未提供のため金額上限判定不可"],
            {"規程": "未提供"},
        )

    limits = cfg.amount_limits
    grade: str | None = getattr(r, "grade", None)

    overs: list[dict] = []       # → NG
    needs_check: list[dict] = [] # → NEEDS_CHECK

    # report に宿泊実態があるか (長距離運転手当の日帰り/宿泊判定)
    has_lodging_cd = any(leg.allowance_cd_lodging for leg in r.legs)

    for leg in r.legs:
        # (1) 出張日当 (手当1CD)
        if leg.allowance_cd_perdiem:
            table = limits.get("出張日当", {})
            _check_grade_allowance(
                leg.leg_no, "出張日当", leg.amount,
                table, grade, overs, needs_check,
            )

        # (2) ホテル代 (手当2CD) — 宿泊実態あり
        if leg.allowance_cd_lodging:
            hotel = limits.get("ホテル代", {})
            _check_hotel(
                leg.leg_no, leg.amount, leg.dest_prefecture, grade,
                hotel, overs, needs_check,
            )

        # (3) 滞在補助費 (手当3CD)
        if leg.allowance_cd_stay:
            table = limits.get("滞在補助費", {})
            _check_grade_allowance(
                leg.leg_no, "滞在補助費", leg.amount,
                table, grade, overs, needs_check,
            )

        # (4) 出張加算日当 (account_name 識別)
        if leg.account_name and "出張加算日当" in leg.account_name:
            table = limits.get("出張加算日当", {})
            _check_grade_allowance(
                leg.leg_no, "出張加算日当", leg.amount,
                table, grade, overs, needs_check,
            )

        # (5) 長距離運転手当 (account_name 識別)
        if leg.account_name and "長距離運転手当" in leg.account_name:
            _check_long_distance(
                leg.leg_no, leg.amount, leg.dest_km_lower,
                has_lodging_cd, limits, overs, needs_check,
            )

        # (6) 車 + 貸借費 分類チェック
        if leg.transport == "車" and leg.account_name and "貸借費" in leg.account_name:
            needs_check.append({
                "明細No": leg.leg_no, "種別": "勘定科目分類",
                "金額": leg.amount,
                "理由": "社用車(旅費交通費)とレンタカー(貸借費)の区分確認が必要",
            })

        # (7) 委託サービス費 + 交通・宿泊系 分類チェック
        _TRAVEL_TRANSPORTS = {"電車･ﾊﾞｽ", "電車・バス", "ﾎﾃﾙ", "ホテル", "車", "車(同乗)",
                              "タクシー", "飛行機", "新幹線", "バス"}
        if (leg.account_name and "委託サービス費" in leg.account_name
                and (leg.transport in _TRAVEL_TRANSPORTS
                     or leg.allowance_cd_perdiem
                     or leg.allowance_cd_lodging
                     or leg.allowance_cd_stay)):
            needs_check.append({
                "明細No": leg.leg_no, "種別": "勘定科目分類",
                "金額": leg.amount,
                "理由": "交通・宿泊費が委託サービス費で申請されています。旅費交通費への変更要確認",
            })

    statuses: list[str] = []
    reasons: list[str] = []
    evidence: dict = {}

    if overs:
        statuses.append(NG)
        reasons.append(f"金額上限超過 ({len(overs)}件)")
        evidence["金額上限超過"] = overs
    if needs_check:
        statuses.append(NEEDS_CHECK)
        reasons.append(f"要確認 ({len(needs_check)}件)")
        evidence["要確認"] = needs_check
    if not overs and not needs_check:
        statuses.append(OK)

    return worst_status(statuses), reasons, evidence


# ---------------------------------------------------------------------------
# 公開エントリポイント
# ---------------------------------------------------------------------------

def check_amount(r: ExpenseReport, cfg: Config) -> CheckResult:
    """金額規程チェック (§6.3). データ整合性 + 規程上限判定の総合."""
    integ_status, integ_reasons, integ_ev = _check_integrity(r)
    rule_status, rule_reasons, rule_ev = _check_rules(r, cfg)

    overall = worst_status([integ_status, rule_status])

    reasons = integ_reasons + rule_reasons
    detail = " / ".join(reasons) if reasons else "金額整合性: 問題なし"

    evidence: dict = {}
    evidence.update(integ_ev)
    evidence.update(rule_ev)
    evidence["整合性判定"] = integ_status
    evidence["規程判定"] = rule_status

    suggestion = ""
    if overall in (NG, NEEDS_CHECK):
        parts: list[str] = []
        if "合計超過" in integ_ev:
            parts.append("明細金額の合計が申告合計を超過しています。金額を確認・修正してください")
        if "差額要確認" in integ_ev:
            parts.append("申告合計と明細合計の差額(手当コードなし)の内訳を確認してください")
        if "行数不一致" in integ_ev:
            parts.append("行数(宣言)と明細件数の差異を確認してください")
        if "明細金額異常" in integ_ev:
            parts.append("金額がマイナス/小計不一致の明細を是正してください")
        if "金額上限超過" in rule_ev:
            parts.append("旅費規定の上限を超過した明細の妥当性を確認してください")
        if "要確認" in rule_ev:
            items = rule_ev["要確認"]
            kinds = list({i.get("種別", "") for i in items if i.get("種別")})
            if kinds:
                parts.append(f"{'・'.join(kinds)}について確認が必要です")
        suggestion = "。".join(parts) + "。" if parts else ""

    return CheckResult(
        status=overall,
        detail=detail,
        evidence=evidence,
        suggestion=suggestion,
    )

File: src js/rules/amount.js
'use strict';

const { OK, NEEDS_CHECK, NG, RULE_MISSING, worstStatus } = require('../models');

function limitFor(limits, key, kmLower) {
  const table = limits[key];
  if (!table || typeof table !== 'object' || kmLower === null || kmLower === undefined) return null;
  let bestKey = null;
  let bestLo  = -1;
  for (const k of Object.keys(table)) {
    const lo = parseInt(k, 10);
    if (!isNaN(lo) && lo <= kmLower && lo > bestLo) { bestLo = lo; bestKey = k; }
  }
  return bestKey !== null ? table[bestKey] : null;
}

function checkIntegrity(report) {
  const reasons = [];
  const evidence = {};
  const statuses = [];

  const gap = report.totalAmount - report.computedTotal;
  const hasAllowance = report.legs.some(l => l.allowanceCdPerdiem || l.allowanceCdLodging || l.allowanceCdStay);

  if (report.computedTotal > report.totalAmount) {
    statuses.push(NG);
    reasons.push(`明細金額合計が申告合計を超過: 計算${report.computedTotal} > 申告${report.totalAmount}`);
    evidence['合計超過'] = { 申告合計: report.totalAmount, 計算合計: report.computedTotal, 超過額: -gap };
  } else if (gap > 0 && !hasAllowance) {
    statuses.push(NEEDS_CHECK);
    reasons.push(`申告合計と明細合計に差額${gap}があるが手当コードなし`);
    evidence['差額要確認'] = { 申告合計: report.totalAmount, 計算合計: report.computedTotal, 差額: gap };
  }

  if (report.actualLegCount !== report.declaredLegCount) {
    statuses.push(NEEDS_CHECK);
    reasons.push(`行数不一致: 宣言${report.declaredLegCount} != 実${report.actualLegCount}`);
    evidence['行数不一致'] = { 宣言行数: report.declaredLegCount, 実明細数: report.actualLegCount };
  }

  const badLegs = report.legs.filter(l => l.amount < 0 || l.subtotal !== l.amount);
  if (badLegs.length > 0) {
    statuses.push(NG);
    reasons.push(`明細金額異常(${badLegs.length}件)`);
    evidence['明細金額異常'] = badLegs.map(l => ({ 明細No: l.legNo, 金額: l.amount, 小計: l.subtotal }));
  }

  return { status: worstStatus(statuses), reasons, evidence };
}

function checkRules(report, cfg) {
  if (!cfg.hasAmountRules()) {
    return { status: RULE_MISSING, reasons: ['旅費規定 未提供のため金額上限判定不可'], evidence: { 規程: '未提供' } };
  }

  const limits = cfg.amountLimits;
  const statuses = [];
  const overs = [];
  const unresolved = [];

  for (const leg of report.legs) {
    if (leg.allowanceCdPerdiem) {
      const cap = limitFor(limits, '日当', leg.destKmLower);
      if (cap === null) {
        unresolved.push({ 明細No: leg.legNo, 種別: '日当', 金額: leg.amount, km下限: leg.destKmLower, 理由: '距離区分未確定' });
      } else if (leg.amount > cap) {
        overs.push({ 明細No: leg.legNo, 種別: '日当', 金額: leg.amount, 上限: cap });
      }
    }
    if (leg.allowanceCdLodging) {
      const cap = limits['宿泊料上限'];
      if (cap === null || cap === undefined) {
        unresolved.push({ 明細No: leg.legNo, 種別: '宿泊料', 金額: leg.amount, 理由: '宿泊料上限未定義' });
      } else if (leg.amount > cap) {
        overs.push({ 明細No: leg.legNo, 種別: '宿泊料', 金額: leg.amount, 上限: cap });
      }
    }
  }

  if (overs.length > 0) { statuses.push(NG); }
  if (unresolved.length > 0) { statuses.push(NEEDS_CHECK); }
  if (overs.length === 0 && unresolved.length === 0) statuses.push(OK);

  const evidence = {};
  if (overs.length > 0) evidence['金額上限超過'] = overs;
  if (unresolved.length > 0) evidence['上限未検証'] = unresolved;

  const reasons = [];
  if (overs.length > 0) reasons.push(`金額上限超過(${overs.length}件)`);
  if (unresolved.length > 0) reasons.push(`上限未検証(${unresolved.length}件)`);

  return { status: worstStatus(statuses), reasons, evidence };
}

function checkAmount(report, cfg) {
  const { status: integStatus, reasons: integReasons, evidence: integEv } = checkIntegrity(report);
  const { status: ruleStatus, reasons: ruleReasons, evidence: ruleEv }   = checkRules(report, cfg);

  const overall = worstStatus([integStatus, ruleStatus]);
  const reasons = [...integReasons, ...ruleReasons];
  const detail  = reasons.length > 0 ? reasons.join(' / ') : '金額整合性: 問題なし';
  const evidence = Object.assign({}, integEv, ruleEv, { 整合性判定: integStatus, 規程判定: ruleStatus });

  let suggestion = '';
  if (overall === NG || overall === NEEDS_CHECK) {
    const parts = [];
    if (integEv['合計超過']) parts.push('明細金額の合計が申告合計を超過しています。金額を確認・修正してください');
    if (integEv['差額要確認']) parts.push('申告合計と明細合計の差額の内訳を確認してください');
    if (ruleEv['金額上限超過']) parts.push('旅費規定の上限を超過した明細の妥当性を確認してください');
    suggestion = parts.join('。') + (parts.length ? '。' : '');
  }

  return { status: overall, detail, evidence, suggestion };
}

module.exports = { checkAmount };
